pkg/virtualgarden/resources/hvpa.yaml

```diff
@@ -0,0 +1,1204 @@
+apiVersion: apiextensions.k8s.io/v1beta1
```

Comment:
I'm really struggling a bit with maintaining this CRD in this repo. I would prefer not having it here and rather returning an error if somebody tries to enable the HVPA feature while the CRD isn't deployed. WDYT?

Reply:
Hard to say for us. We do not have such a good overview about all the components of the gardener as you. Is there any other component which should be responsible to deploy this CRD? If not, this has to be done manually before the installation of the virtual garden. Or do we need some component installing all prerequisites such as this CRD?

Reply:
The HVPA controller is deployed by the gardenlet when the runtime cluster is registered as a seed. The gardenlet does also deploy the HVPA CRD in the same course.
Having the HVPA CRD again in landscape-setup was only done to allow the creation of the `Hvpa` resource for the initial deployment, i.e., when the virtual garden is deployed (seed registration only happens later). 
Personally, I don't like this duplication and would rather vote for only allowing the enablement of the HVPA feature after the seed registration has happened (or after the HVPA components were deployed by other means).

cc @amshuman-kr

Reply:
ping @amshuman-kr

Reply:
I do not have the full context here. Is this dependency for deploying HVPA CRD for both the below scenarios?

1. On the compute cluster on which the virtual garden would be installed. I.e. the cluster that would be registered on the virtual garden cluster as a soil cluster
2. On all the other seed clusters

> would rather vote for only allowing the enablement of the HVPA feature after the seed registration has happened

If we rely on gardenlet to install HVPA, does it mean that we will do the virtual garden installation in two steps? Once without HVPA and then enable HVPA for the virtual garden after the soil cluster has been registered as a seed (which would install HVPA)?

> (or after the HVPA components were deployed by other means).

@rfranzke  Do you have any options in mind?

PS: I would guess we might have to address the same questions if we want to put the virtual-garden ETCD also under the control of `etcd-druid` (which, we ideally should).

Reply:
> Is this dependency for deploying HVPA CRD for both the below scenarios?

Generally, HVPA on seed clusters is deployed by gardenlet as part of the seed reconciliation. Hence, if at all, only 1. is relevant.

> If we rely on gardenlet to install HVPA, does it mean that we will do the virtual garden installation in two steps? Once without HVPA and then enable HVPA for the virtual garden after the soil cluster has been registered as a seed (which would install HVPA)?

Yes, this would be the consequence, but probably cleaner solution compared to duplicating the HVPA deployment which will always conflict with the deployment of gardenlet.

> Do you have any options in mind?

For this and also the etcd-druid case, the only idea I have is allowing to disable HVPA/etcd-druid per seed and deploy both via virtual-garden. Though, I don't like it too much because it means continuous and duplicate maintenance effort (HVPA/etcd-druid must be maintained in g/g and g/v-g).

Reply:
@rfranzke , @amshuman-kr, looks as if we have a similar problem with VerticalPodAutoscaler ([see](https://sap-ti.slack.com/archives/G0170EXLULE/p1626791853015700?thread_ts=1626783682.014100&cid=G0170EXLULE)). This is already used in virtual garden but the crd is deployed in the gardenlet.

I do not understand much of the general setup and nothing about druid but I would vote for another (basic) component installing technical prerequisites such as some CRDs, HVPA controller etc. to prevent cyclic dependencies or more than one installation run to activate a feature.

Reply:
Hmm... I don't like the duplication either. Could it be possible to avoid the duplication by separating the installation of HVPA and VPA (CRDs and the resources instances for the objects in `garden` namespace) as an extension? However, this approach may not work for etcd-druid if we want to use it to back the virtual garden ETCD (unless there is another way to deploy the extension that is independent of gardener-controller-manager and gardenlet).

Reply:
> This is already used in virtual garden but the crd is deployed in the gardenlet.

Maybe the code can be adapted such that the VPA resources are only deployed if the GVK is found in the system?

> but I would vote for another (basic) component installing technical prerequisites such as some CRDs, HVPA controller etc. to prevent cyclic dependencies. More than one installation run to activate a feature makes setup more complicated.

> Could it be possible to avoid the duplication by separating the installation of HVPA and VPA (CRDs and the resources instances for the objects in garden namespace) as an extension? 

Both approaches result in duplication. Right now, I don't see another acceptable way than having multiple runs to activate certain features.

Reply:
Sorry for observing this so late.

Where is HVPA required:
- as Rafael said: only when setting up the virtual Garden with hvpa feature [here](https://github.com/gardener/gardener/blob/dd7584c9396e1b8cf7c1b5351edc092a55d35375/charts/gardener/controlplane/charts/runtime/templates/apiserver/hvpa.yaml). Otherwise it will be deployed by the Gardenlet during the Seed setup.


However, currently there are 3 places where hvpa is installed

- The gardenlet
- In the landscape setup virtual-garden component
- in the landscape setup gardenlet component

Rafaels suggestion: remove HVPA installation from landscape-setup completely. Disallow deploying the virtual-garden helm chart with hvpa feature if HVPA is not installed.

Advantage: 
- no code duplication between Gardenlet and landscape-setup / landscaper

Disadvantage (in my opinion)
- less streamlined installation for open source consumers (2 installation runs required so that the Gardenlet can bootstrap the Seed cluster first)

Approach @achimweigel and what I previously talked with @schrodit about: externalize HVPA & VPA in a dedicated module to at least remove duplication between Gardenlet landscape-setup component and the virtual-garden component. 

I would agree with Rafael here.  Mainly because this would save us from creating another dedicated component for now. We can just document that 2-step installation behavior if hvpa is desired.
That will save us some time that can be invested in finishing other landscaper components.

Reply:
Hi @danielfoehrKn , @rfranzke ,

we have changed this PR as you proposed. The HVPA CRD is no longer created. We just check that it exists if it is needed.